### DISCUSSION TITLE
Change JSON parser lib

### DISCUSSION THREAD

**Discussion by williamthome** [zotonic/zotonic#3340] [Ideas]
Hi!

JSON parser is used a lot in Zotonic.
The current Zotonic parser is `jsxrecord`.
I benchmarked some JSON parser libs, and there is a huge difference comparing the result:

```erlang
% The result is average iterations per second (is low because the JSON file is 8.1 MB).
% Higher is best.
run(decode, Filename) ->
    {ok, Bin} = file:read_file(Filename),
    ebench:run([
        {thoas, decode, [Bin]},    % result: 10
        {jsx, decode, [Bin]},      % result:  7
        {jiffy, decode, [Bin]},    % result: 46
        {jsone, decode, [Bin]},    % result: 14
        {jsxrecord, decode, [Bin]} % result:  7 <- Zotonic parser
    ], #{samples => 20});
run(encode, Filename) ->
    {ok, Bin} = file:read_file(Filename),
    ebench:run([
        {thoas, encode, [ begin {ok, JSON} = thoas:decode(Bin)
                                , JSON end ]},       % result: 10
        {jsx, encode, [jsx:decode(Bin)]},            % result:  6
        {jiffy, encode, [jiffy:decode(Bin)]},        % result: 30
        {jsone, encode, [jsone:decode(Bin)]},        % result: 13
        {jsxrecord, encode, [jsxrecord:decode(Bin)]} % result:  7 <- Zotonic parser
    ], #{samples => 20}).
```
`jiffy` is really fast.
I have not tested changing the parser in Zotonic, just noticed that improvement is easy since Zotonic encapsulates the JSON parser in the `z_json` module.

See the [bench repo](https://github.com/williamthome/ebench/blob/main/src/ebench_json.erl) for more info or a local test.

## Libs URL
- [erlperf](https://hex.pm/packages/erlperf) (the bench lib)
- [thoas](https://hex.pm/packages/thoas)
- [jsx](https://hex.pm/packages/jsx)
- [jiffy](https://hex.pm/packages/jiffy)
- [jsone](https://github.com/sile/jsone)
- [jsxrecord](https://github.com/zotonic/jsxrecord)

## My setup
- CPU: Intel(R) Core(TM) i7-7700HQ CPU @ 2.80GHz
- OS: {unix, linux}
- Erlang: Erlang/OTP 25 [erts-13.2] [source] [64-bit] [smp:8:8] [ds:8:8:10] [async-threads:1] [jit:ns]

**Comment by mworrell** [comment]
We are using jsxrecord as a wrapper.  Maybe we should have called it jsonrecord instead.

jsxrecord "understands" the Erlang records for serialization and deserialization, and maps `null` to `undefined`.
Under the hood it just calls jsx (with some pre- and postprocessing).

I am open to swap the JSON library used by jsxrecord.

From what I remember jiffy is very fast, but (being a NIF) had a problem with locking the scheduler, so not scaling nicely for parallel work. Might have been fixed though.

**Reply by williamthome** [reply]
Ah, right, I have not looked at the `jsxrecord` code.
I will investigate and try to do some tests with `jiffy`.
Any news I'll come back.

**Comment by mworrell** [comment]
The JSON encoding library is now `euneus`
